### DISCUSSION TITLE
Preventing download when using -w %{header_json}

### DISCUSSION THREAD

**Discussion by God-damnit-all** [curl/curl#9167] [Q&A] [answered]
(Note: I'm using PowerShell 7.3 on Windows 10 in case that's relevant to the solution.)

I'm trying to get json-formatted headers of a file into a variable. Unfortunately, trying to bind the command to a variable, piping it, or putting it in a subexpression (even if I try putting it in `cmd /c`) causes it to start initiating a file transfer, which isn't good because it's a very large file and the purpose of checking the headers is to see if it's been updated.

The one workaround I've found is doing `$head = (curl.exe -sSLI -w '%{stderr}%{header_json}' $url >$null) 2>&1` which discards stdout and then redirects stderr to stdout. However, this isn't a great solution in my opinion, more of a hacky workaround that takes away a useful output stream.

I've tried `--range 0-0` but it seems that parameter doesn't work at all with following redirects no matter what the parameter's argument is. `--max-filesize 1` (0 seems to be 'no limit') is just as hacky since it outputs a 'Maximum file size exceeded' error every time, and that'll interfere with PowerShell's experimental 'native command error throwing' feature, which I do like to use.

I could theoretically run it in a job and then get the job's output, though again, that makes my use of the 'native command error throwing' feature worthless.

I've looked throughout all of `--help all` and I've done multiple searches, but I can't seem to find an adequate solution. Please help.

_(I'm fine with any `--libcurl` solutions if need be.)_

**Comment by jzakrzewski** [comment]
As far as I know, the `-I` sends a HEAD request, which is supposed to be answered WITHOUT body and I thought curl would discard the body anyway if there was any...

**Reply by God-damnit-all** [reply]
> As far as I know, the `-I` sends a HEAD request, which is supposed to be answered WITHOUT body

Yes, but it's not given in json format. Furthermore, if there are redirects, it posts headers for every single one.

> I thought curl would discard the body anyway if there was any...

Apparently not.

**Reply by bagder** [reply]
Can you show us a command line that uses `-I` in which curl shows a returned body?

**Reply by God-damnit-all** [reply]
> Can you show us a command line that uses `-I` in which curl shows a returned body?

You misunderstand. `-I` does not return a body. The problem is that it's occupying stdout along with the json-formatted header information from `-w`, so I have to somehow filter it out.

**Reply by bagder** [reply]
Your "Preventing download" should thus be interpreted as "I want to prevent the tool from writing anything to stdout", right?

On a Linux machine I would do this:
~~~
curl -I localhost -w '%{header_json}' -so /dev//null
~~~

Does Windows have anything like `/dev/null` for `-o` ?

**Reply by God-damnit-all** [reply]
> Your "Preventing download" should thus be interpreted as "I want to prevent the tool from writing anything to stdout", right?
> 
> On a Linux machine I would do this:
> 
> ```
> curl -I localhost -w '%{header_json}' -so /dev//null
> ```
> 
> Does Windows have anything like `/dev/null` for `-o` ?

The problem is that regardless of it writing to stdout, it still downloads the body. So if said body is a 2GB file I want the headers for, that presents a problem.

**Reply by bagder** [reply]
No, with `-I` curl does not download any body.

**Reply by God-damnit-all** [reply]
> No, with `-I` curl does not download any body.

I overlooked that part since I tend to group the parameters together.

Yes, this works with $null, thank you.

**Comment by God-damnit-all** [comment, marked as answer]
Answer from @bagder:
> Your "Preventing download" should thus be interpreted as "I want to prevent the tool from writing anything to stdout", right?
> 
> On a Linux machine I would do this:
> 
> ```
> curl -I localhost -w '%{header_json}' -so /dev//null
> ```
> 
> Does Windows have anything like `/dev/null` for `-o` ?

This with NUL instead of /dev/null, according to https://stackoverflow.com/a/313118
